README.md
# civ6save analysis

## Setup

Install [ImHex](https://imhex.werwolv.net/).

## Dump decompressed binary file

I use the `dump_decompressed` component from the `civ6_pipeline` to produce the decompressed binary data file.

```bash
viash run ../civ6_pipeline/src/civ6_save_renderer/dump_decompressed/config.vsh.yaml -- \
  --input saves/000085.Civ6Save \
  --output saves/000085_map.bin
```

## Open project in ImHex

When opening the decompressed binary file in ImHex, the first thing to do is search for the array
`0E 00 00 00 0F 00 00 00 06 00 00 00`. In the civ6_pipeline, this is called the "SearchBuffer".

At SearchBuffer + 0x10, the first tile of the map is found. I added bookmarks to the ImHex project
to be able to easily distinguish the first couple of tiles. At the edge of the map, these tiles typically
begin with `00 00 01 00 00 01 00 00`.

![](images/imhex1.png)

## Variable tile length

The reason for the tearing observed in [viash-io/civ6_pipeline#9](https://github.com/viash-io/civ6_pipeline/issues/9)
is because the tile information is variable in length. Up until now, I came across tiles that are 55, 72, 79, 96, 99, or 116 in length.

* Standard tile information: 55 bytes
* If the tile affects CO2 in some way: +24 bytes
  - In addition, if the tile is a ski resort or tunnel: +20 bytes
* When the flag next to the CO2 is set (I don't know the meaning of this flag yet): +44 bytes
* If the tile is owned by a player: +17 bytes

## Hex pattern

I'm trying to create a hex pattern with ImHex to be able to make sense of the Civ6Save file more easily. By loading it into the Pattern Editor and evaluating the code in the Console, you can explore the tile information quite easily.

![](images/imhex2.png)

## Pattern Data

Note that the values observed in the Pattern Data should match up with the values in the [Civ6Save Data Sheet](https://docs.google.com/spreadsheets/d/1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs/edit) document.

![](images/imhex3.png)



testmap/CHANGELOG.md
# testmap v0.4

- All terrains (tiles 0 - 16)
- All features (tiles 74 - 94)
- All continents (tiles 296 - 338)
- Most improvements (tiles 1481 - 1539, all except barbarian outpost)
- All resources (tiles 815 - 862)
- All world wonders (scattered throughout the map)
- All roads (tiles 1633 - 1642)
- All units (tiles 2813 - 2955)
- 10 players, 1 city state (Brussels), 12 cities
- Multiple districts
- Feature rotations (tiles 2295 - 2301)
- Coastal height (tiles 2306 - 2309)
- Resource counts (tiles 963 - 966)
- Different river orientations (see sheet)
- Different cliff orientations (see sheet)

Specific player properties:
- America (Player 0) has all tiles revealed
- Arabia (Player 1) has access to all civics
- Australia (Player 2) has access to all techs
- Aztec (Player 3) has 123465+x gold
- Babylon (Player 4) has 456789+x faith

# testmap v0.2

- All terrains (tiles 0 - 16)
- All features (tiles 74 - 94)
- Most continents (tiles 296 - 338)
- Most improvements (tiles 1481 - 1539)
- Most resources (tiles 816 - 862)
- All world wonders (scattered throughout the map)
- All roads (tiles 1633 - 1642)
- 6 players, 1 city state (Brussels), 9 cities

Specific player properties:
- America (Player 0) has all tiles revealed
- Arabia (Player 1) has access to all civics
- Australia (Player 2) has access to all techs
- Aztec (Player 3) has 123465+x gold
- Babylon (Player 4) has 456789+x faith


testmap/process.R
library(tidyverse)

options(width = 200)

as_hex <- function(x) x %>% as.integer %>% as.hexmode %>% format %>% paste(collapse = "")
width <- 74

map <- jsonlite::read_json("testmap/testmap_v0.4.json", simplifyVector = TRUE) %>%
  .$tiles %>%
  as_tibble() %>%
  mutate(
    i = row_number() - 1,
    x = i %% width,
    y = floor(i / width),
    # buffer0_len = map_int(buffer0, length),
    # buffer1_len = map_int(buffer1, length),
    # buffer0_chr = map_chr(buffer0, as_hex),
    # buffer1_chr = map_chr(buffer1, as_hex)
  ) %>%
  select(i, x, y, everything())

map %>% filter(grepl("\\?\\?\\?", terrain)) %>% select(1:10)

library(googlesheets4)

gs4_auth("robrecht.cannoodt@gmail.com")

terrains <- read_sheet("1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "terrains")
terrains$terrain_int32le <- map$terrain[terrains$testmap_tile + 1]
write_sheet(terrains, "1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "terrains")

features <- read_sheet("1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "features")
features$feature_int32le <- map$feature[features$testmap_tile + 1]
write_sheet(features, "1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "features")

improvements <- read_sheet("1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "improvements")
improvements$improvement_int32le <- map$improvement[improvements$testmap_tile + 1]
write_sheet(improvements, "1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "improvements")

continents <- read_sheet("1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "continents")
continents$continent_int32le <- map$continent[continents$testmap_tile + 1]
write_sheet(continents, "1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "continents")

resources <- read_sheet("1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "resources")
resources$resource_int32le <- map$resource[resources$testmap_tile + 1]
write_sheet(resources, "1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "resources")

roads <- read_sheet("1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "roads")
roads$road_int32le <- map$road[roads$testmap_tile + 1]
write_sheet(roads, "1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "roads")


# check rivers
map[c(665, 666, 741, 742, 2748, 2750, 2752)+1,] %>% as.data.frame

# check cliffs
map[c(665, 668, 670, 671, 2682, 2684, 2686)+1,] %>% as.data.frame

# check non-pillaged vs pillaged improvements
map[c(1481, 1482, 1484, 1407, 1408, 1410)+1,] %>% as.data.frame

# check feature rotations
map[c(2295:2301)+1,] %>% as.data.frame

# coastal lowland
map[c(2306:2309)+1,] %>% as.data.frame

# units
map[c(2812:2816)+1,] %>% as.data.frame

# resource counts
map[c(963:966)+1,] %>% as.data.frame

# districts
map[c(2004:2007, 2079:2081, 2153)+1,] %>% as.data.frame


# generate enums
terrains <- read_sheet("1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "terrains")
features <- read_sheet("1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "features")
improvements <- read_sheet("1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "improvements")
continents <- read_sheet("1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "continents")
resources <- read_sheet("1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "resources")
roads <- read_sheet("1bOlgW25zpWOUTPcPcNDbfpXK5f90J2BBaBuzwypABQs", "roads")


fun <- function(df, nameenum, namecol, hexcol) {
  names <- gsub("[^a-zA-Z]", "", stringi::stri_trans_general(df[[namecol]], id = "Latin-ASCII"))
  hexes <- map_chr(str_extract_all(df[[hexcol]], ".."), function(x) paste(rev(x), collapse = ""))
  codes <- paste(paste0("    ", names, " = 0x", hexes), collapse = ",\n")

  paste0("enum ", nameenum, " : s32 {\n", codes, "\n};\n")
}
cat(
  fun(terrains, "Terrain", "terrain_name", "terrain_hex"),
  fun(features, "Feature", "feature_name", "feature_hex"),
  fun(improvements, "Improvement", "improvement_name", "improvement_hex"),
  fun(continents, "Continent", "continent_name", "continent_hex"),
  fun(resources, "Resource", "resource_name", "resource_hex"),
  sep = "\n"
)



# width <- 74
# df <- map %>%
#   mutate(
#     i = row_number() - 1,
#     x = i %% width,
#     y = floor(i / width),
#   ) %>%
#   civ6saves:::add_coordinates()

# df

# ggplot(df) +
#   geom_point(aes(x, y, colour = terrain))


# xy_ratio <- 0.5 / cos(30/180*pi)


# plot_terrain_map <- function(tib, alpha = 1) {

#   plot_empty_map(tib) +
#     ggforce::geom_regon(aes(fill = terrain), alpha = alpha) +
#     ggforce::geom_regon(aes(fill = feature, r = xy_ratio * .9), data = tib %>% filter(feature == "FeatureType::Ice"), alpha = alpha) +
#     geom_text(aes(label = "^"), tib %>% filter(terrain == "FeatureType::Hill"), alpha = alpha) +
#     geom_text(aes(label = "^"), tib %>% filter(terrain == "FeatureType::Mountain"), fontface = "bold", alpha = alpha)#+
#     #geom_segment(aes(x = xa, xend = xb, y = ya, yend = yb), colour = feature_palette[["River"]], rivers, size = 1, alpha = alpha) +
#     #scale_fill_manual(values = c(terrain_palette, feature_palette))
# }

# plot_terrain_map(df)

# plot_empty_map <- function(tib) {
#   ggplot(tib, aes(x = x0, y = y0, x0 = x0, y0 = y0, sides = 6, angle = 30 / 180 * pi, r = xy_ratio)) +
#     theme_void() +
#     theme(axis.line=element_blank(),
#           axis.ticks=element_blank(),
#           axis.title=element_blank(),
#           axis.text.y=element_blank(),
#           axis.text.x=element_blank(),
#           panel.border = element_blank(),
#           panel.grid.major = element_blank(),
#           panel.grid.minor = element_blank(),
#           panel.background = element_blank(),
#           plot.title = element_text(hjust = .5, size = 16),
#           plot.subtitle = element_text(hjust = .5, size = 12),
#           legend.position = "bottom") +
#     coord_equal(expand = FALSE)
# }

# #' @importFrom rlang %|%
# #' @export
# plot_terrain_map <- function(tib, alpha = 1) {
#   rivers <- get_river_coordinates(tib)

#   plot_empty_map(tib) +
#     ggforce::geom_regon(aes(fill = terrain_name), alpha = alpha) +
#     ggforce::geom_regon(aes(fill = feature_name, r = xy_ratio * .9), data = tib %>% filter(feature_name == "Ice"), alpha = alpha) +
#     geom_text(aes(label = "^"), tib %>% filter(terrain_form == "Hill"), alpha = alpha) +
#     geom_text(aes(label = "^"), tib %>% filter(terrain_form == "Mountain"), fontface = "bold", alpha = alpha) +
#     geom_segment(aes(x = xa, xend = xb, y = ya, yend = yb), colour = feature_palette[["River"]], rivers, size = 1, alpha = alpha) +
#     scale_fill_manual(values = c(terrain_palette, feature_palette))
# }

# #' @export
# plot_feature_map <- function(tib) {
#   plot_terrain_map(tib, 0.3) +
#     geom_point(aes(colour = feature_name), size = point_size, tib %>% filter(!is.na(feature_name)), shape = 20) +
#     scale_colour_manual(values = feature_palette)
# }

# #' @export
# pal_many_categories <- rep(c(
#   RColorBrewer::brewer.pal(9, "Set1"),
#   RColorBrewer::brewer.pal(8, "Set2"),
#   RColorBrewer::brewer.pal(12, "Set3"),
#   RColorBrewer::brewer.pal(8, "Pastel1"),
#   RColorBrewer::brewer.pal(8, "Pastel2"),
#   RColorBrewer::brewer.pal(8, "Dark2")
# ), 100)

# #' @export
# pal_some_categories <- c(
#   RColorBrewer::brewer.pal(8, "Dark2"),
#   RColorBrewer::brewer.pal(8, "Pastel2"),
#   RColorBrewer::brewer.pal(8, "Set2")
# )